pytorch_lightning/core/lightning.py

```diff
@@ -1714,75 +1560,123 @@ def get_tqdm_dict(self) -> Dict[str, Union[int, str]]:
-            return {}
-
-
-def _collect_init_args(frame, path_args: list) -> list:
-    """
-    Recursively collects the arguments passed to the child constructors in the inheritance tree.
-
-    Args:
-        frame: the current stack frame
-        path_args: a list of dictionaries containing the constructor args in all parent classes
-
-    Return:
-          A list of dictionaries where each dictionary contains the arguments passed to the
-          constructor at that level. The last entry corresponds to the constructor call of the
-          most specific class in the hierarchy.
-    """
-    _, _, _, local_vars = inspect.getargvalues(frame)
-    if '__class__' in local_vars:
-        cls = local_vars['__class__']
-        spec = inspect.getfullargspec(cls.__init__)
-        init_parameters = inspect.signature(cls.__init__).parameters
-        self_identifier = spec.args[0]        # "self" unless user renames it (always first arg)
-        varargs_identifier = spec.varargs     # by convention this is named "*args"
-        kwargs_identifier = spec.varkw        # by convention this is named "**kwargs"
-        exclude_argnames = (
-            varargs_identifier, kwargs_identifier, self_identifier, '__class__', 'frame', 'frame_args'
-        )
-
-        # only collect variables that appear in the signature
-        local_args = {k: local_vars[k] for k in init_parameters.keys()}
-        local_args.update(local_args.get(kwargs_identifier, {}))
-        local_args = {k: v for k, v in local_args.items() if k not in exclude_argnames}
+        if not frame:
+            frame = inspect.currentframe().f_back
+        init_args = get_init_args(frame)
+        assert init_args, 'failed to inspect the self init'
+        if not args:
+            hp = init_args
+            self._hparams_name = 'kwargs' if hp else None
+        else:
+            isx_non_str = [i for i, arg in enumerate(args) if not isinstance(arg, str)]
+            if len(isx_non_str) == 1:
+                hp = args[isx_non_str[0]]
+                cand_names = [k for k, v in init_args.items() if v == hp]
+                self._hparams_name = cand_names[0] if cand_names else None
+            else:
+                hp = {arg: init_args[arg] for arg in args if isinstance(arg, str)}
+                self._hparams_name = 'kwargs'
+
+        # `hparams` are expected here
+        if hp:
+            self._set_hparams(hp)
+
+    def _set_hparams(self, hp: Union[dict, Namespace, Any]) -> None:
+        if isinstance(hp, Namespace):
+            hp = vars(hp)
+        if isinstance(hp, dict):
+            hp = AttributeDict(hp)
+        if isinstance(hp, PRIMITIVE_TYPES):
+            raise ValueError(f'Primitives {PRIMITIVE_TYPES} are not allowed.')
+        if not isinstance(hp, ALLOWED_CONFIG_TYPES):
+            raise ValueError(f'Unsupported config type of {type(hp)}.')
```

Comment:
I suggest having a distinct code path for each input type. It's much easier to understand:
```python
if isinstance(hp, Namespace):
  ...
elif isinstance(hp, dict):
  ...
elif isinstance(hp, ALLOWED_CONFIG_TYPES):
  ...
else:
  raise ValueError(f'Unsupported config type of {type(hp)}.')
```

Reply:
almost the second `if` has to stay simple but the est is better :]